Subfield VOLE

lzjluzijie opened the pull request:
Implements #122

- [x] Noisy VOLE
- [x] PPRF
- [x] Coding
- [x] Silent VOLE
- [x] u128 is done 
- [x] array<u32, 4> 
- [x] any integer array
- [x] u64: cannot use SilentBaseType::BaseExtend
- [x] real fields?
- [x] better TypeTrait
- [x] clean up
- [x] Malicious
- [x] Merge new pprf
- [x] Unify implementations (implement non-template pprf,vole,ot etc using template)

lzjluzijie commented:
For https://github.com/osu-crypto/libOTe/blob/master/libOTe/Tools/SilentPprf.cpp#L524-L525, we need to generate random elements from `pprf.mBaseOTs` (using AES)?

ladnir commented on libOTe/Vole/Subfield/NoisyVoleReceiver.h:
```diff
@@ -0,0 +1,107 @@
+    msg.resize(otMsg.size(), y.size());
+
+    for (u64 ii = 0; ii < (u64)otMsg.size(); ++ii) {
+      prng.SetSeed(otMsg[ii][0], z.size() * sizeof(F) / sizeof(block)); // todo
```
seems you are assuming `sizeof(F)` will be a multiple of `block`.

lzjluzijie replied:
Use `fromBlock` instead?

ladnir replied:
`fromBlock` would work.

ladnir commented on libOTe/Vole/Subfield/NoisyVoleReceiver.h:
```diff
@@ -0,0 +1,107 @@
+    for (u64 ii = 0; ii < (u64)otMsg.size(); ++ii) {
+      prng.SetSeed(otMsg[ii][0], z.size() * sizeof(F) / sizeof(block)); // todo
+      auto& buffer = prng.mBuffer;
+      F *buf = (F *)buffer.data();
```
Maybe ok for not but not very safe. You are assuming you can just cast like this. Maybe `static_assert(std::is_trivial<F>)` or something similar.

ladnir commented on libOTe/Vole/Subfield/NoisyVoleReceiver.h:
```diff
@@ -0,0 +1,107 @@
+        z[j] = z[j] + buf[j];
+
+        F twoPowI = 0;
+        *BitIterator((u8*)&twoPowI, ii) = 1;
```
again, you are assuming this is the correct thing. More generally there should be a function that the user can give you for doing this. Ideally this should work when the input isn't a POD datatype. For now this is ok.

ladnir commented on libOTe_Tests/SilentOT_Tests.cpp:
```diff
@@ -928,6 +929,220 @@ void Tools_Pprf_test(const CLP& cmd)
+//     }
+// };
+
+using u128 = unsigned __int128;
```
not portable.

lzjluzijie replied:
The reason I use this is if I use the class version commented above, I would get error

```
In file included from /home/halulu/dev/libOTe/out/install/linux/include/coproto/Socket/Socket.h:20,
                 from /home/halulu/dev/libOTe/out/install/linux/include/coproto/coproto.h:11,
                 from /home/halulu/dev/libOTe/libOTe_Tests/../libOTe/Tools/Coproto.h:9,
                 from /home/halulu/dev/libOTe/libOTe_Tests/../libOTe/Tools/SilentPprf.h:20,
                 from /home/halulu/dev/libOTe/libOTe_Tests/SilentOT_Tests.cpp:3:
/home/halulu/dev/libOTe/out/install/linux/include/coproto/Proto/Buffers.h: In instantiation of ‘macoro::coroutine_handle<void> coproto::internal::MoveSendProto<Container>::await_suspend(const H&) [with H = macoro::coroutine_handle<macoro::detail::task_promise<void, true> >; Container = std::vector<std::array<u128, 4>, std::allocator<std::array<u128, 4> > >]’:
/home/halulu/dev/libOTe/out/install/linux/include/macoro/coro_frame.h:152:29:   required from ‘auto macoro::await_suspend(Awaiter&, coroutine_handle<T>, enable_if_t<((! has_void_await_suspend<Awaiter, coroutine_handle<T> >::value) && (! has_bool_
await_suspend<Awaiter, coroutine_handle<T> >::value)), empty_state>) [with Awaiter = coproto::internal::MoveSendProto<std::vector<std::array<u128, 4>, std::allocator<std::array<u128, 4> > > >; T = detail::task_promise<void, true>; enable_if_t<((! has_void_await_suspend<Awaiter, coroutine_handle<T> >::value) && (! has_bool_await_suspend<Awaiter, coroutine_handle<T> >::value)), empty_state> = empty_state]’
/home/halulu/dev/libOTe/libOTe_Tests/../libOTe/Tools/SubfieldPprf.h:621:25:   required from ‘macoro::task<void> osuCrypto::SilentSubfieldPprfSender<G, F>::Expander::run() [with G = u128; F = u128]’
/home/halulu/dev/libOTe/libOTe_Tests/../libOTe/Tools/SubfieldPprf.h:300:75:   required from ‘macoro::task<void> osuCrypto::SilentSubfieldPprfSender<G, F>::expand(osuCrypto::Socket&, osuCrypto::span<const T>, osuCrypto::PRNG&, osuCrypto::MatrixView<F>, osuCrypto::PprfOutputFormat, bool, osuCrypto::u64) [with G = u128; F = u128; osuCrypto::Socket = coproto::Socket; osuCrypto::span<const T> = nonstd::span_lite::span<const u128, 18446744073709551615>; osuCrypto::u64 = long unsigned int]’    
/home/halulu/dev/libOTe/libOTe_Tests/SilentOT_Tests.cpp:1104:28:   required from here
/home/halulu/dev/libOTe/out/install/linux/include/coproto/Proto/Buffers.h:260:53: error: no matching function for call to ‘asSpan(std::vector<std::array<u128, 4>, std::allocator<std::array<u128, 4> > >&)’
  260 |                                 if (internal::asSpan(mContainer).size() == 0)
      |                                     ~~~~~~~~~~~~~~~~^~~~~~~~~~~~
compilation terminated due to -Wfatal-errors.
```

It seems chl cannot send it directly, how should I fix this?

ladnir replied:
so if you send data, then the thing `T` you're sending either needs to be `std::is_trivial<T>::value == true` or you can send something that looks like a vector of trivials things, e.g.  `std::vector<T>` where `T` is trivial. Specifically, when you pass something to coproto, it checks this using https://github.com/Visa-Research/coproto/blob/main/coproto/Common/TypeTraits.h#L160 

If this trait is true for your type, then that function `internal::asSpan(mContainer)` exists here https://github.com/Visa-Research/coproto/blob/main/coproto/Proto/Operation.h#L122

Maybe I should make this error message a bit more descriptive ;)

ladnir commented on libOTe/Tools/SubfieldPprf.h:
```diff
@@ -0,0 +1,1437 @@
+    {
+    public:
+        u64 mDomain = 0, mDepth = 0, mPntCount = 0;
+        std::vector<G> mValue;
```
I think this should be the full field F

ladnir commented on libOTe/Tools/SubfieldPprf.h:
```diff
@@ -0,0 +1,1437 @@
+        Matrix<std::array<block, 2>> mBaseOTs;
+        
+        std::function<void(u64 treeIdx, span<AlignedArray<F, 8>>)> mOutputFn;
+        std::function<F (const block& b)> fromBlock;
```
this will have very bad performance. Its ok for now but at some point we will need to switch to having the caller pass in a `type trait` object that defines `F,G`, and all of these related functions. That way these function calls can be inlined.

ladnir commented on libOTe/Tools/SubfieldPprf.h:
```diff
@@ -0,0 +1,1437 @@
+                                // where each half defines one of the children.
+                                aes[keep].hashBlocks<8>(parent.data(), child.data());
+
+                                if (d < pprf.mDepth - 1) {
```
a nested if statement like this might have an impact on performance. Could be work checking

ladnir reviewed (commented):
Looks like a great start

ladnir commented:
do not use QuasiCyclicCode. We can just use `ExConv`. When you are ready I can give some pointers on how.

ladnir commented:
As mentioned in the comments, It best to thing of `F,G` as some arbitrary type that you can assume nothing about. Whenever you need some functionality out of them, you should achieve this by a function call.

I would suggest changing the type level template to be some `FieldTraits` which has subtypes `FieldTraits::F,FieldTraits::G`. Whenever you want some functionality, you can call the function `FieldTraits::functionality(...)`.

lzjluzijie commented:
Thank you for your review, for `ExConv`, I should read https://eprint.iacr.org/2023/882 ?

ladnir commented:
yeah, likely would be helpful. You don't need to understand the proof but the basic idea. So the intro and the start of section 5

lzjluzijie commented:
I wonder if we should define some virtual class `Field` that supports `operator+`, `operator*`, and `fromBlock`? I am not expert in cpp. I would implement this "interface" style if in other languages, but is this a good choice in cpp(might be slow than template)?

ladnir commented:
virtual would be the natural way of doing it but it has very bad performance. 

The cost of a function call is orders of magnitude higher than say `+`.  C++ often allows you to avoid this cost by having the compiler inline the function, essentially copy pastes the function body into the call sight so there is no function call. However, this can not work for virtual functions because the compiler does not know which "virtual" instances is currently being used.

You can still achieve this same type of "virtual" behavior with templates. In fact, templates are basically virtual functions where you are additionally telling the compiler exactly which type is being used.


So concretely, my suggestion is the following:

- The caller will be responsible for defining a type traits class
```
struct TypeTrait
{
    using F = u128;
    using G = u8;

    F sampleF(PRNG& prng);
    F fromBlock(const block& blk);
    F getPowerOfTwo(u64 power);
    ...
};
```

- each of the subfield classes will take a `TypeTrait` as a template parameter. You can then define everything else in terms of that.
- `TypeTrait` will essentially form the "virtual" class that other languages use. However, the caller can define their `TypeTrait` to have these functions be `inline` and therefore the overhead of the function calls will disappear.
- The caller is responsible for correctly implementing this "interface". There is no "base class" to inherit from. The caller is just expected to know what function to implement based on the documentation. If a function is missing, then the code doesn't compile.

make sense?

An example of this pattern can be found here. https://github.com/Visa-Research/volepsi/blob/main/volePSI/PxUtil.h#L550 In this case I called it `Helper` instead of `TypeTraits` and it has some other abstractions. But the same basic idea. It was used to allow the caller to call the OKVS for VOLE-PSI with basically any type as long as they provided the `Helper`. In particular, they could even call it when the "element", in out case `F` was some variable vector.

lzjluzijie commented:
Update: I fixed some bugs in PPRF, and the u128 silent VOLE is working if comment out the final coding part. The code part is still not working, and I will continue working on it tomorrow. The code seems a little complicated, but I think it might work by changing some `^` to `+`. No need to review code at this time, but you could take a look at `ExConvCode_encode_u128_test` if my test make sense if you have time. Thank you.

ladnir commented:
That's right, basically just need to use plus. Encode2 is for performance. It just encodes twice. You can just call the single in code twice if you want.

lzjluzijie commented:
I got silent VOLE test passed with F=G=u128. The ExConvCode is indeed well templated and I only changed a few `^` to `+`. I will test it with other fields later.

As for make this more generic, I wonder why the current block does not support operator*, and operator+ seems adding left and right parts independently as u64. Shall we change it to behave like GF(2^128) elements, or we define some `class F128: block` and over write the operators?

ladnir commented:
`block` is defined that way due to its primary purpose being an abstract on the `__m128i` built in type. This type models special cpu instructions (aka intriniscs). It does not have 128 bit addition or multiple, so `block` doesn't. There is 2x 64bit addition as you mention. Maybe I shouldn't have overloaded operator + since it's a bit confusing. But I did so oh well ha.

ladnir commented:
As suggested above, we shouldn't even use `operator+` or `operator^`. Instead there should be a type traits object that handles all the operations for you. You simply passing in the arguments. But this can happen at then end once the basic version is working

lzjluzijie commented:
So we should do something like

```cpp
struct TypeTrait
{
    using F = u128;
    using G = u8;
    ...
    inline F Add(const F& a, const F& b)
    ...
};
```

and I was suggesting

```cpp

struct F128: block {
  inline F128 operator+(const F128& rhs) const {
    ...
  }
};

struct TypeTrait
{
    using F = F128;
    using G = F128;
    ...
};
```

(not using virtual). Is there performance difference?

ladnir commented:
Yes, it will be about 10x faster. Maybe in some cases the compiler might be smart enough to remove this overhead but best not to relay on the compiler that way.

ladnir commented:
In simples example the compiler does this thing called devirtualization which basically can inline virtually functions. But there's no guarantee that it will happen. However, if you use inline functions, you are basically guaranteed there's no cost. The function will disappear.

lzjluzijie commented:
I see, I will change the code. And for ExConv, I should also modify it to use inline member function, not operator `+` or `^`?

lzjluzijie commented:
One more question please: In my second example, F128 inherits block, but the operator function is also inline without virtual. Is this kind of inline different from the member function inline in the first example? If so, is it because of F128 inherits block?

ladnir commented:
Ok, i took a closer look at you example. Inheritance != virtual. You can inherit all you want. The functions can still be inline and they would behave the same as if you wrote new version of them. What has a performance overhead is when you use the `virtual` keyword on a function and then override it in some base class. Specifically, consider the example
```
struct base {
  virtual void foo() = 0;
};
struct derived0 : base {
void foo() override { std::cout << "derived0 << std::endl; }
};
struct derived1 : base {
void foo() override { std::cout << "derived1 << std::endl; }
};
```
When someone has a `base* v;` and they call `v->foo();` what happens is that at the begining of `derived0, derived1`, there is a hidden pointer that points to something that called a v-table. This v-Table looks like:
```
struct vTableForDerived0{
      void(* baseFooFunctionPointer)(void* derived0);
};
struct vTableForDerived1{
      void(* baseFooFunctionPointer)(void* derived1);
};
```
and the derived classes actually look like
```
struct derived0 : base {
void* pointerToVTable;
void fooImpl() { std::cout << "derived0 << std::endl; }
};
struct derived1 : base {
void* pointerToVTable;
void fooImpl() { std::cout << "derived1 << std::endl; }
};
```
Now the `v->foo();` call gets transformed into
```
auto* vTable = v->pointerToVTable;
vTable->baseFooFunctionPointer(v); // this calls derived0::fooImpl() or derived1::fooImple() 
```
so now it might be a bit clearer why doing virtual is a bit of a performance hit. You have to chase the pointer to the vtable and then you have to call `baseFooFunctionPointer`. So say instead of printing something we were just doing a `+` on a `u64`. Then you are doing all this, plus the function call which itself has to do like a bunch of stuff (flush variables our of registers, update the program counter, make space for a new stack frame, execute the call instruction, and then you do a simple `+`. 

In some cases the compiler is smart and can look at your program and say "hay, even though `v` is a `Base*`, i know it's actually `derived0` and therefore i can just call `derived::fooImpl`. I will even inline it". However, if there are many possible derived types and the compiler can't see which is the type being used, then it can't do this devirtualization. 

In your example above you don't use virtual and therefore none of this stuff happens. They are just normal function calls. As you can see in your assembly, these get inlined (no function calls made).

------------------------------------
Correct, you can define `add` like you did above and use that wherever you want to do a plus. Additionally, I would suggest not requiring that the `TypeTrait` object only have a static function. So these subfield protocols will take the normal `F,G` values an input and an instance of `TypeTraits`. You use this instance to do all the additions and everything else.

Consider the case where you want to allow the modulus to be many different values. If `add` is static, then this modulus must be somewhere within the `F` type. e.g.
```
struct F {
  u64 val;
  u64 modulus;
};
```
So now all `F`s have to be twice as big. However, if `TypeTraits` is an actual object, you can simply stick the modulus in it. Now `F` is only a `u64` in the example above.

The alternative would be for `F` to have `modulus` hardcoded into the "type". But this is a bit less flexible. Although maybe its good enough haha.

lzjluzijie commented:
Thank you very much for your explanation. I understand that virtual is slow. Now we should have two approaches

The first is use operators with field elements(no virtual), for example

```cpp
// Definition
struct F128 {
    block b;

    inline F128 operator*(const F128& lhs, const F128& rhs) const {
        block r = lhs.b.gf128Mul(rhs.b);
        return F128(r);
    }
};

struct TypeTrait
{
    using F = F128;
    using G = F128;
    ...
   // no explict Add needed.
};

// Usage
```

The second approach:

```cpp
struct TypeTrait
{
    using F = F128;
    using G = F128;
    ...
    inline F Add(const F& a, const F& b)
    ...
};
```

If my understanding is correct: both approaches should be similar performance, given there is no virtual and inlines are used. I think one advantage of the first approach that it needs less code. If we want to use primitive type for these protocol, we don't need to write `Add` functions in the second approach. However, the second approach might be safer. In the first case, users might have some unsafe pointer operation that make `F*` to `block*` and use wrong operator function. But I think inside libOTe we can make sure this never happen and it should be fine?

As for the modulus, I suggest that we can make F templated with modulus, and I think this is more convenient(in the first approach).

So I personally prefers the first approach. Again I don't know much about cpp, is there any issue with it? Thank you.

ladnir commented:
Use, you can do the first if you want. I might change it later though. Regarding the user, you can give a default version of TypeTrait to assumes F, G are normal types with +, etc.

lzjluzijie commented:
Makes sense. So you think the second is better? I can do that directly.

ladnir commented:
I'd prefer the second because you are separating the data (F, G) from the operations (add, sample, etc). You can use the same datatype (eg block) for both a 128 bit prime field and GF128. But it's fine either way.

lzjluzijie commented:
Makes sense. I will do more tests with larger fields and then do it.

lzjluzijie commented:
I got some issue with `genSilentBaseOts`. If I configure it with `SilentBaseType::Base` and run BaseOT, it works normally. However, if I use `SilentBaseType::BaseExtend` with NoisyVOLE requiring 64 not 128 OTs, assertion in `libOTe/TwoChooseOne/SoftSpokenOT/SoftSpokenMalOtExt.cpp:56` would fail.

I'm not familiar with SoftSpokenOT and I wonder if there is any special reason to use it, not something simpler like IKNP?

ladnir commented:
For Semi-honest iknp would be fine. For simplicity I use soft-spoken. It sends a bit less data as well...

Where is it failing exactly? I suspect is soft spoken is failing then iknp would too. This have the same api. Hard to say though.

For silent vole that's sounds right. For noisy vole, you're saying G=gf(32) and F=gf(k*32) doesn't work? The issue isn't clear to me.

lzjluzijie commented:
SoftSpokenMalOtExt.cpp:56 would fail. I am saying G is uint32, F is array of uint32, not fields.

ladnir commented:
So the receiver has `a in G^n`, the sender has `X in F`.

The receiver computes `vi = a * F(2^i)`. The user can define `*` however they want, more or less. The receiver acts as OT sender
```
mi0 = Ri
mi1 = Ri + vi
```
where `Ri in F^n is random. The sender learns 
```
mi_{Xi} = Ri + Xi *  F(2^i) * a
```
When you sum the received messages you get 
```
B = sum_i mi_{Xi} 
  = sum_i Ri + Xi *  F(2^i) * a
  = sum_i Ri + sum_i Xi *  F(2^i) * a
  = C + X * a
```
where `C = sum_i Ri`.

You want to define `G` as the `Z_{2^32}` group and `F` as the vector space over it. This is fine. It should work.

ladnir commented:
For softspoken,im not sure. Seem not right. What command/test produces the error. I can run it.

lzjluzijie commented:
Sorry was very busy, now back to this.

- Vector of u32 works using the noisy method. I made a mistake in the power function of vector. Now the silent vole work fine with vector of u32 of size 4.
- For the softspoken error, it happen when `SET(CMAKE_BUILD_TYPE Debug)`, and works fine in Release mode. I guess it might be some memory not properly initialize, but I didn't figure out the issue. You can test it with `-u 123`.
- Should I move the implementation of ExConv from cpp files to headers? The instantiation files are quite big and inconvenient to use now.

I will do more testing with larger "fields" and clean up code later.

ladnir commented:
Yeah, change the code however makes sense. The instantiation thing was to improve compile times but I don't think it makes sense anymore. I can look into soft spoken

lzjluzijie commented on libOTe/Vole/Subfield/Subfield.h:
```diff
@@ -0,0 +1,127 @@
+  using G = T;
+
+  static OC_FORCEINLINE F fromBlock(const block &b) {
+    PRNG prng(b, N * sizeof(T));
```
Making a new PRNG every time has low performance.

We could use some fixed PRNG (make fromBlock nonstatic and use a member prng or something) object and set different seed. Is there any better way?

ladnir replied:
- if `sizeof(F)<= sizeof(block)`, you can just `memcpy` `b` into the return value.
- if `sizeof(F) > sizeof(block)`, you can use `oc::mFixedKeyAes::hashBlock(b + block(i,i))` to generate additional blocks and then `memcpy` these into the return value.

lzjluzijie commented on libOTe/Vole/Subfield/NoisyVoleReceiver.h:
```diff
@@ -0,0 +1,104 @@
+    if (z.size() == 0) throw RTE_LOC;
+
+    memset(z.data(), 0, sizeof(F) * z.size());
+    msg.resize(otMsg.size(), z.size());
```
When F=G^m with large m, this would eat all memory, given `z.size()` is already 400 from silent VOLE. We don't need a matrix and I think it's here for chl performance? Any idea to improve the performance?

lzjluzijie replied:
Each element of `msg` is in F, and `otMsg` has the bit length of `F`, so this communication(and memory now) is m^2 complexity.

lzjluzijie replied:
Suppose G has k bits, F=G^m, and length of the noisy subfield VOLE (numPartitions) be `n`. Currently, for bit of the delta in F, the OT message has `(2*) n` elements in F, which is `(2*) n*k*m` bits each. Since F has `k*m` bits, the overall communication complexity is `n*k^2*m^2`.

Alternatively, we do `m` times noisy VOLE with F=G. Each time the communication complexity is `O(n*k^2)`, so overall is `O(n*m*k^2)`. I assume the security is the same. We should be able to use template to check if F!=G(which means F=G^m for m > 1), and use this approach.

What do you think?

ladnir replied:
**Current Design:** We want to compute `a * X + B =  C` for `a in G^n`, `B,C in F^n` and `X in F`, i.e. `X=delta`. Let `k = log2(|G|), t=km = log2(|F|)`. 

The number of OTs is `t`. not `n`. We do an OT for each bit of `X=delta`.

The overhead of the protocol is sending `t` elements of `F^n` That is, for each of the `t` bits of `X`, we do an OT, and then send a single vector in `F^n`. Therefore `t^2 * n = k^2 * m^2 *n` bits in total.

**Alternative Design:** There is an alternative design where you do an OT for each bit of `a`. You would then need `kn` OTs. This will be a lot more in most cases.

Are you suggesting this?

**Multi-threading ?:** You could but we we only do very small noisyVoles, so I dont think it makes sense to make the noisy vole itself parallel. I'd think if you wanted to do that, you could just have the caller make multiple noisy voles and call those in parallel.

lzjluzijie replied:
No, I am suggesting doing `m` times of noisy G-G VOLE, each VOLE has `k` OT messages, so it's still `m*k` messages. But since OT messages are sending elements of G now, not F, so message size for each OT message is reduced from `n*k*m` to `n*m`.

lzjluzijie replied:
I agree multi-threading would not make a huge improvement. I just wanted to say it's possible.

ladnir replied:
ah, right, so when we compute 
```
vi = a * F(2^i)
```
as described above, you could naively represent this as a `F^n` vector but you're saying that's kinda wasteful because you can also just represent it as 
```
vi = a  G(2^{i mod k}) 
```
The sender can then scale these up on their side by multiplying by `F(2^{i - (i mod k)})`. Yeah, makes sense. nice.

lzjluzijie replied:
Yeah, I'm going to implement this.

ladnir replied:
I guess you could also change how this works in that you dont need "pre" multiply with `2^i` at all. Instead, you could have
```
vi = a
```
do OT on 
```
mi0 = Ri
mi1 = Ri + vi = Ri + a
```
and then have both parties scale `mi0,mi1` up by `F(2^i)`

lzjluzijie replied:
This means both end need to do a multiplication, but know only the receiver need to do a multiplication.

Also, I'm not sure about the security of my approach mentioned above. We need the receiver to input the same y for `m` times. I think it should be secure though.

ladnir replied:
I guess in both cases you need both parties to do the scaling. Either way is fine.

Security should be fine. You only need Semi-honest security and privacy against malicious parties. We don't require correctness. So changing the `a=y` is fine.

lzjluzijie reviewed (commented):
I made the silent VOLE work for G=u32/u64 and F=G^m for large m, with some performance issue.

lzjluzijie commented:
There is already lots of changes in this PR. I think I should put that Noisy VOLE in a separate PR.

Also I noticed that, many functions in AES.h like `finalEnc`, `roundEnc` are not force inlined, and are indeed not inlined when I test it(at least in DEBUG mode). I can write another PR for this, and fix the header issues in cryptoTools.

ladnir commented:
Do not use debug mode to make decisions about what being inlined. You'd need to use release. In debug builds the compiler often prefers not to inline because it can make debugging harder.

Sure, multiple PRs is good if that's what you prefer. Either way is fine with me.

lzjluzijie commented:
Makes sense, but is there any reason to use the normal `inline`, not `OC_FORCEINLINE` in AES.h?

I added a `F128` type which is essentially a block, but with the proper `+`, `-`, and `*` in GF(2^128). I added a test for it in ExConv. But there are lots of place needed to change from block to F128. What should I do?

ladnir commented:
So lance roy is the one who added force inline. Not totally sure if there was a real need for it. Inline tells the compiler please inline but the compiler can still not inline while force tells the compiler to always do it. Typical if it's a good idea to inline then the compiler will do the right thing with just inline. Sometime not though. 

If you are going to use force inline you should have a benchmark that shows it actually helps. In some cases it might hurt (the code becomes bigger).


For exconv, it's not clear to me what the issue is. In general you should replace block with F. But if I recall it's already templated so I'm confused about the question.

lzjluzijie commented:
OK, that makes sense.

I am saying that there are other places, such as VOLE test, needs to replace block with F. Maybe do that in another PR as well?

ladnir commented:
leave the existing tests. Make new ones and adapt them lightly. So for the block ones, + and - should both be mapped to xor.

lzjluzijie commented:
I added a new namespace `osuCrypto::Subfield` and put all new copies in that namespace so the changes won't affect existing tests. The code is working now and you can take a look. Malicious security is not implemented yet and the optimization for noisy I mentioned would be another PR after this.

ladnir commented:
i think you didn't commit your tests 
```
C:\Users\peter\repo\libOTe\out\build\x64-Debug\libOTe\libOTe_Tests.lib(UnitTests.cpp.obj) : error LNK2001: unresolved external symbol "void __cdecl osuCrypto::Subfield::Subfield_ExConvCode_encode_test(class osuCrypto::CLP const &)" (?Subfield_ExConvCode_encode_test@Subfield@osuCrypto@@YAXAEBVCLP@2@@Z)
C:\Users\peter\repo\libOTe\out\build\x64-Debug\libOTe\libOTe_Tests.lib(UnitTests.cpp.obj) : error LNK2001: unresolved external symbol "void __cdecl osuCrypto::Subfield::Subfield_Tools_Pprf_test(class osuCrypto::CLP const &)" (?Subfield_Tools_Pprf_test@Subfield@osuCrypto@@YAXAEBVCLP@2@@Z)
C:\Users\peter\repo\libOTe\out\build\x64-Debug\libOTe\libOTe_Tests.lib(UnitTests.cpp.obj) : error LNK2001: unresolved external symbol "void __cdecl osuCrypto::Subfield::Subfield_Noisy_Vole_test(class osuCrypto::CLP const &)" (?Subfield_Noisy_Vole_test@Subfield@osuCrypto@@YAXAEBVCLP@2@@Z)
C:\Users\peter\repo\libOTe\out\build\x64-Debug\libOTe\libOTe_Tests.lib(UnitTests.cpp.obj) : error LNK2001: unresolved external symbol "void __cdecl osuCrypto::Subfield::Subfield_Silent_Vole_test(class osuCrypto::CLP const &)" (?Subfield_Silent_Vole_test@Subfield@osuCrypto@@YAXAEBVCLP@2@@Z)
C:\Users\peter\repo\libOTe\out\build\x64-Debug\libOTe\frontend\frontend_libOTe.exe : fatal error LNK1120: 4 unresolved externals
  ninja: build stopped: subcommand failed.

```

ladnir commented:
pushed some changes to my subfield branch.  Below are my comments

- i found a bug in my ExConv does where i call `_mm_laod_ps` for types that aren't `block`. This can fail if the thing being loaded isn't 16 byte aligned. I think in release build this code gets optimized away in these cases (the result isn't used) and so doesn't crash. But on debug linux builds I'm getting a crash. I fixed this by adding an if statement to guard the call.
- `F128` is not portable. Like i mentioned before, you can not use `__int128`. It is not standard c++. You need to use standard c++ or optionally enable platform-specific features behind `#ifdef`macros. 
- never include `"cryptoTools/cryptoTools/..."` or `"out/..."`. These will not work correctly when you install libOTe. 
- it would be nice if you used the same indenting as the rest of the library. No worries if this isn't easy to set.
- You make many assumptions about the `F,G` types. ideally, we make no assumptions apart from the function that `TypeTrait`s provides. Things like using `sizeof(F)`, setting and reading bits using `BitIterator`, assuming `F,G` have member functions for `+, *`.
- you can give a "default" impl of `TypeTraint<F,G>` that just calls your preferred methods that assume that `F,G` are POD types and have `+,*` members. But in general, for example, it would be nice if the impl also worked if say `F,G` are just pointers to their underlying values. Of, in the case of `block`, we want `+` to map to `^`. 
- In the end we want to replace the existing impls of `ExConv, Vole` with your new one. So things should be general & efficient enough to handle both cases with no additional overhead. For the general case, we can "hide" the impls for `QuasiCyclic` and the other codes.  That way, we only need a general impl for `ExConv`.
- Also, for `SubfieldExConvConfigure(...)`, the keyword `static` (for a "free function") is not what you should use when you give the implementation in the header. Instead, use `inline`. Both work but `static` has other meanings. For class member functions, `static` means something different and you are using it correctly.

ladnir reviewed (commented):
See comment.

lzjluzijie commented:
I see, I will change to use functions from `TypeTrait` instead of using `F`, `G` directly. What is your configuration for indentation?

ladnir commented:
4 spaces. no single line if statement (can't place breakpoints).

lzjluzijie commented:
I made changes to the TypeTraits and it works fine except for the encoder. I'm not sure if you want the encoder take the full TypeTrait(which contains two types) of VOLE as template, since we may want to use the encoder independently.

Also can you take a look at the SoftSpoken issue I mentioned above? It's still error in DEBUG mode(but is fine in release mode).

ladnir commented:
The encoder can take it as a template. Later we can revisit the decision if needed.

I tested soft spoken on my machine and it worked... What compiler are you using?

lzjluzijie commented:
I changed the ExConv to use TypeTrait.

I am using `gcc (Debian 12.2.0-14) 12.2.0`. Can you try `-u 131` on my new commit?

ladnir commented:
ok, the issue was just with the assert. the code is fine.

ladnir commented:
do all the unit tests pass for you? seems there is an issue with silent vole. see the CI above

lzjluzijie commented:
Yes, it's an error with dualEncode2, fixing it now.

lzjluzijie commented:
It's quite strange. It seems commit [5132392](https://github.com/osu-crypto/libOTe/pull/127/commits/5132392ac29532aae925eed9be69f46b8a9b23de) changed the result of  `accumulate` for blocks. I tried to fix it for an hour but was not successful, will try again later.

lzjluzijie commented:
OK I figured it out.  There is some issue with the block optimized accumulate. I commented out all `std::is_same<block, T>::value` code and it's working now. I'm not sure what is the optimization doing and `b[0].get<i32>` seems strange to me. I guess we can further improve this part of code?

ladnir commented:
No worries, I can take a look at some point.

ladnir commented:
i refactored pprf so you'll likely want to merge those. The new one is a bit simpler.

lzjluzijie commented:
Sure, I will take a look later 👍

lzjluzijie commented:
I rebased and squashed my commits to the new master, the old commits are in https://github.com/lzjluzijie/libOTe/tree/subfield-old

I made some progress now, and I want to confirm that you removed the "last level". I have two solutions to process the last level for our template type `F`. One approach is adding `lastLevel` back as before, another approach is call `TypeTraits::fromBlock` to convert block to `F` when needed. Which one do you prefer?

ladnir commented:
Last is still there I just made it one big buffer that's shares with the other layers. This simplified some issues elsewhere. So you should just be able to make the buffer bigger or smaller so the "last layer" is the right size. Make sense?

ladnir commented:
https://github.com/osu-crypto/libOTe/blob/master/libOTe/Tools/SilentPprf.cpp#L502

Hopefully you just need to update this and then change the type. Tried to make it not difficult to integrate.

lzjluzijie commented:
I think that `last` is the `lastOT`, which should be type `F` (change buffer type from block to u8?). Additionally, we still need https://github.com/osu-crypto/libOTe/blob/master/libOTe/Tools/SilentPprf.cpp#L859 for lastLevel to copyOut?

ladnir commented:
Correct, change buffer from vec<block> to vec<u8> and update LastType to be array<F,4> or similar.

For the last layer, it's most the same as before I think. We perform the extra logic to convert from block to F and then copy out accordingly. For the special case of F=block (or similar) and output format = interleaved, there an optimization where the last layer is already the output. Ideally this should remain. However, when that doesn't make sense, ge F isn't a block, then we will have to copy/construct the Fs from the last layer blocks.

lzjluzijie commented:
I see, I will try to add back `lastLevel` and push a draft version for review

ladnir commented:
I've also started implementing silent Fp triple generation which will require this. Will be nice to finally have. https://eprint.iacr.org/2022/1035

ladnir commented:
> I see, I will try to add back `lastLevel` and push a draft version for review

Did you do it a different way?

ladnir commented:
I see what you did. Hmm, yeah not sure what's best. Maybe your way is fine too?

lzjluzijie commented:
Yeah I think so

lzjluzijie commented:
I merged the new pprf, and I think the code is ready for review now.

I think it's good to have unified implementation. However, it needs a lot of work, and I guess most people would use the block non-subfield version. So I think we can merge this separate subfield code and then unify them in a later PR.

ladnir commented:
Sure, I can't do it well. Not a big deal. Should I take a close look then?

lzjluzijie commented:
Yes please review it

lzjluzijie commented:
For malicious, I think we can also do it in a separate PR. As for the optimized subfield noisy VOLE, I had already implemented it months ago, and I can open a new PR once this is merged.

ladnir reviewed (commented):
ok, so looking at the impl I think we can do `TypeTraits` better.

For example, 
```cpp
    template<typename TypeTrait>
    class SilentSubfieldVoleSender;
```
could be 
```cpp
    template<
        typename F, 
        typename G = F, 
        typename TypeTrait = DefaultTrait<F, G>
        >
    class SilentSubfieldVoleSender;
```
This way the caller doesn't need to know about the trait stuff unless they want to do something special. 

Then, for the pprf's `gAes`. its best not to put that in a header. every file that include it will get its own independent copy of `gAes`. If there wasn't an anonymous namespace around it you would have gotten a linker error (duplicate definitions). 

For the ExConv code, I dont think the class itself need to be a template. You can just pass the type trait stuff to the member functions, like how you are now passing `F,G`.

Finally, at some point I want to have a instance of `TypeTrait` that the functions can take as a runtime parameter. That way it can store whatever runtime info is needed to add and multiply, e.g. maybe you want `Vec` to be variable length but you dont want to have each `Vec` redundantly and inefficiently storing its own size.

Also, as a style thing, if the class is complicate and implementing in the header, its best to still just declare the function in the class def and then below the class implement each function (as a template).

In `sampleBaseVoleValues`, you assume you can `memcpy` into a vector of `G`'s instead of using from block. Also, its important that these are non-zero. For `block` this is basically just assumed but for smaller types we need to set it. As a test, you could try to instantiate this with a non-trivial `F,G` and this i think this wouldn't compile.

We should fix this last one for sure. If you want to do the other great but if not we can merge and I'll work on them later.

ladnir commented:
hmm, I kinda would like to see everything unified before merging into master. 

For malicious, we can start with conditionally allowing it if `F==block`. But i think we can do something reasonable in the general case.

Let me know what you want to do. I can unify if you prefer.

lzjluzijie commented:
Yeah it makes sense to have this new TypeTrait to unify subfield.

lzjluzijie commented:
For `gAes`, I don't know how to put it in non-header file while our implementions using it is in header files.

For ExConv, yes we can put TypeTrait in member functions, but what is the advantages to do this?

ladnir commented:
For `gAes`,  you would need to find a cpp file for it to live and remove the anonymous namespace. Then in the header you can declare it as `extern`, i.e.
```
    // A public PRF/PRG that we will use for deriving the GGM tree.
    extern const std::array<AES, 2> gAes;
```
`extern` means it exists somewhere but you don't know where until linking. 

Yeah. for compiles time the issue is that every file cpp you include the impl recompile the whole thing. This compounded by the fact that i think ExConv takes awhile. To fix this we will need to split the template into a header "A" where is declared and a separate (less used) header "B" where its implemented. In some cpp file, you can then explicitly instantiate the versions of subfield that you want. This cpp file will include "B" and then instantiate the classes you want. These will be compiled only once in that single cpp file. I did this for the old ExConv code.

What the advantage of making ExConv not a template? it makes it so that you don't need to compile as many things over and over again. Also, there are several functions that take the ExConv as parameter that you had to turn into a template only to support ExConv. By making it not a template all these also become not a template.

lzjluzijie commented:
I see, thank you. I will try my best to implement them

ladnir commented:
For G=integer rings, we will also need to be careful. It's been shown that rings (as opposed to fields) get less security. One can fix this by doubling the noise rate. However, that 2x the communication overhead. An alternative heuristic approach is to require all of the noise elements to be odd.

So I think we should be able to query TypeTrait about what type of field/ring is it, also, how many bits is G. We have the following cases:

1. If it's binary, then we simply set the noise element to 1 and skip noisy vole.
2. If it's a ring, we sample a random odd element (or double the noise but that's less efficient)
3. If it's a field, we samole a random nonzero element.

See https://eprint.iacr.org/2022/712

lzjluzijie commented:
I updated `gAes`, `ExConv`, and `sampleBaseVoleValues`. I added `DefaultTrait` and we can now use `SilentSubfieldVoleReceiver<block>` directly, and it should be ready to unify. I'm not sure about styling and the exact way you want to unify, can you take it(and combine malicious for block at the same time)?

As for the ring/linearMix issue, it makes sense to me. Since no one is using it now, I think we can do it in a separate PR.

ladnir commented:
@lzjluzijie  thanks for the PR.  ended up refactoring lots of things (not just your code) but its all merged now.

ladnir commented:
lets see, we have `log2 |G| = k, F = G^m` and want to do a vole of size `n`. We currently have

1. `km` OTs, each requiring `s` bits of communication to generate (for softspoken `s=64` or `s=32`, for base ot `s=512`, for silent `s=1`).
2. For each OT, we send a derandomization string of size `kmn`
3. Overall we will have `km * (km n + s)` bits of communication.

You are proposing to switch the direction:

1. `nk` OTs each with `s` comm.
2. For each OT, we send a derandomization string of size `km`.
3. Overall we will have `kn * (km + s)`.

So for `m=1,k=128, n=100` & softspoken `s=32`, the old way requires `128^2 *100 + 128^2 * 32=2,162,688` bits, the new way requires `128^2 *100 + 128 * 100 = 1,576,800`. If `n>128`, then the new way is worse. Current LPN parameters has `n=170` or so but you could argue thats too high, some say `n=64` is fine or even less.

For `m=4, k=32, n=100`, the old way requires `128 * 128 * 100 + 128 * 32=1,642,496`. The new way requires `32*100 (128 + 32)=512,000`.

This actually has a second (larger?) benefit, the rounds complexity of vole goes down to 2.
```
Silent Sender         Silent Receiver
             base ot 1 
      ( pprf & c choice bits)
     <------------------------

              base ot 2,
            noisy vole msg
      ------------------------>
```
In the old way we had to do the `pprf` OT and `c` OT in different directions. This means the best you could do was 3 rounds (with base Ots)! Now we can even use OT extension to get 3 rounds and avoid doing a PK operation for each.
